### PR TITLE
Feature/improvements of sdl dataset

### PR DESCRIPTION
**Pull request by Geheiner**
### What changes are included in the pull request?
Adding additional features to ScalaClassSparkDsNTo1Transformer after using it in a real-time environnement.

### CODE REVIEW

File: sdl-core/src/test/scala/io/smartdatalake/workflow/action/ScalaClassSparkDsNTo1TransformerTest.scala [hunk 1]
```diff
@@ -287,6 +292,47 @@ class ScalaClassSparkDsNTo1TransformerTest extends FunSuite with BeforeAndAfter
...
+      getClass.getResource("/configScalaClassSparkDsNto1Transformer/usingDataObjectIdWithPartitionAutoSelect.conf").getPath)),
+      partitionValues =
+        Some(Seq(PartitionValues(
+          Map(("year" -> "1992"),
+            ("month" -> "04"),
+            ("day" -> "25")
+          ))))
+    )
+    //Run SDLB
+    val (subFeeds, _): (Seq[SubFeed], Map[RuntimeEventState, Int]) = sdlb.startSimulationWithConfigFile(sdlConfig, Seq(srcDO1, srcDO2))(session)
```
Comment (Geheiner) [inline sdl-core/src/test/scala/io/smartdatalake/workflow/action/ScalaClassSparkDsNTo1TransformerTest.scala:325]: interestingly, even though startSimulationWithConfigFile should only execute init-phase, I'm able to test it with real data. Now Idea why

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/spark/transformer/ScalaClassSparkDsNTo1Transformer.scala [hunk 2]
```diff
@@ -114,11 +122,33 @@ case class ScalaClassSparkDsNTo1Transformer(override val name: String = "ScalaCl
...
+      }
+    }
+    else {
+      resWithSelect
+    }
+  }
+
+  private def getCaseClassColumnNames(className: String): Seq[String] = {
+    val outputCaseClass = Environment.classLoader.loadClass(className)
+    outputCaseClass.getDeclaredFields.toSeq.map(_.getName)
```
Comment (zzeekk) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/action/spark/transformer/ScalaClassSparkDsNTo1Transformer.scala:153]: Clean reflection implementation would be
```
import scala.reflect.runtime.universe._
def classAccessors[T: TypeTag]: List[MethodSymbol] = typeOf[T].members.collect {
  case m: MethodSymbol if m.isCaseAccessor => m
}.toList
```
We should add this to ProductUtils, as we need this in several places.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/spark/transformer/ScalaClassSparkDsNTo1Transformer.scala [hunk 3]
```diff
@@ -134,8 +164,15 @@ case class ScalaClassSparkDsNTo1Transformer(override val name: String = "ScalaCl
       case (param, paramIndex) if param.typeSignature <:< typeOf[Dataset[_]] =>
         val paramName = param.name.toString
         val dsType = param.typeSignature.typeArgs.head
-        val df = tolerantGet(dfs, paramName).getOrElse(throw new IllegalStateException(s"($actionId) [transformers.$name] DataFrame for DataObject $paramName not found in input DataFrames"))
-        val ds = EncoderUtil.createDataset(df, dsType)
+        val df = tolerantGet(dfs, paramName).getOrElse(throw new IllegalStateException(s"($actionId) [transformers.$name] DataFrame for DataObject $paramName not found in input DataFrames: ${dfs.keys.mkString(",")}"))
```
Comment (zzeekk) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/action/spark/transformer/ScalaClassSparkDsNTo1Transformer.scala]:
```suggestion
        val df = tolerantGet(dfs, paramName)
          .orElse(tolerantGet(dfs, paramName.stripPrefix("ds")))
          .getOrElse(throw new IllegalStateException(s"($actionId) [transformers.$name] DataFrame for DataObject $paramName not found in input DataFrames: ${dfs.keys.mkString(",")}"))
```

File: sdl-core/src/test/scala/io/smartdatalake/workflow/action/ScalaClassSparkDsNTo1TransformerTest.scala (hunk 1 above)
Comment (zzeekk) [reply sdl-core/src/test/scala/io/smartdatalake/workflow/action/ScalaClassSparkDsNTo1TransformerTest.scala:325]: exec Phase includes writing DataObjects, thats why simulation is using init. But in fact there are several special rules for simulation in init phase... see usage of ActionPipelineContext.simulation.

**Review by zzeekk — APPROVED** [review]
Looks good, i have only some small additions

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/spark/transformer/ScalaClassSparkDsNTo1Transformer.scala (hunk 2 above)
Comment (Geheiner) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/action/spark/transformer/ScalaClassSparkDsNTo1Transformer.scala:153]: Done, but in a diffent pr: https://github.com/smart-data-lake/smart-data-lake/pull/552

**Pull request merged**